tensornetwork/contractors/opt_einsum_paths/path_contractors.py

```diff
@@ -16,15 +16,77 @@
 
 import functools
 import opt_einsum
+import tensornetwork as tn
 from tensornetwork.network import TensorNetwork
-from tensornetwork.network_components import Edge
+from tensornetwork.network_components import Edge, BaseNode
 from tensornetwork.contractors.opt_einsum_paths import utils
-from typing import Any, Optional, Sequence
+from typing import Any, Optional, Sequence, Iterable, Union
 
+#TODO: add return types of functions back once TensorNetwork is gone
 
-def base(net: TensorNetwork,
-         algorithm: utils.Algorithm,
-         output_edge_order: Optional[Sequence[Edge]] = None) -> TensorNetwork:
+
+def _base_nodes(nodes: Iterable[BaseNode],
+                algorithm: utils.Algorithm,
+                output_edge_order: Optional[Sequence[Edge]] = None) -> BaseNode:
+  """Base method for all `opt_einsum` contractors.
+
+  Args:
+    nodes: A collection of connected nodes.
+    algorithm: `opt_einsum` contraction method to use.
+    output_edge_order: An optional list of edges. Edges of the
+      final node in `nodes_set`
+      are reordered into `output_edge_order`;
+      if final node has more than one edge,
+      `output_edge_order` must be pronvided.
+
+  Returns:
+    Final node after full contraction.
+  """
+  nodes_set = set(nodes)
+  tn.check_connected(nodes_set)
+  edges = tn.get_all_nondangling(nodes_set)
+  #output edge order has to be determinded before any contraction
+  #(edges are refreshed after contractions)
+  if output_edge_order is None:
```

Comment:
Also, why are we not raising an error when the `output_edge_order` isn't given and the number of output edge's is > 1?

Setting it to be the difference of two sets gives it a non-deterministic order and can cause really subtle bugs.

Reply:
that's a good point. I think we should throw an error. I'll fix this